src/coreclr/jit/llvm.cpp

```diff
@@ -1643,47 +1405,298 @@ llvm::DILocation* createDebugFunctionAndDiLocation(struct DebugMetadata debugMet
+            case GT_STORE_LCL_VAR:
+                indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_STORE_OBJ : GT_STOREIND;
+                storedValue = node->AsOp()->gtGetOp1();
+                break;
+            case GT_LCL_VAR:
+                indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_OBJ : GT_IND;
+                break;
+            default:
+                unreached();
+        }
+        node->ChangeOper(indirOper);
+        node->AsIndir()->SetAddr(lclAddress);
+        if (GenTree::OperIsStore(indirOper))
+        {
+            node->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+            node->AsOp()->gtOp2 = storedValue;
+        }
+
+        CurrentRange().InsertBefore(node, offset);
+        CurrentRange().InsertAfter(offset, shadowStackVar);
+        CurrentRange().InsertAfter(shadowStackVar, lclAddress);
+    }
+}
+
+void Llvm::ConvertShadowStackLocals()
+{
+    _shadowStackLclNum = _compiler->lvaGrabTemp(true DEBUGARG("shadowstack"));
+    LclVarDsc* shadowStackVarDsc = _compiler->lvaGetDesc(_shadowStackLclNum);
+    shadowStackVarDsc->lvIsParam = 1;
+    shadowStackVarDsc->lvType = var_types::TYP_I_IMPL;
+
+    for (BasicBlock* _currentBlock : _compiler->Blocks())
+    {
+        _currentRange = &LIR::AsRange(_currentBlock);
+        for (GenTree* node : CurrentRange())
+        {
+            if (node->OperIs(GT_STORE_LCL_VAR, GT_LCL_VAR))
+            {
+                ConvertShadowStackLocalNode(node->AsLclVarCommon());
+            }
+            else if (node->IsCall())
+            {
+                GenTreeCall* callNode = node->AsCall();
+
+                if (callNode->gtCallType == CT_HELPER || callNode->gtReturnType == var_types::TYP_VOID)
+                {
+                    // helper calls are built differently
+                    continue;
+                }
+
+                // we can't do these yet
+                if (callNode->gtCallType == CT_INDIRECT || _isRuntimeImport(_thisPtr, callNode->gtCallMethHnd))
+                {
+                    failFunctionCompilation();
+                }
+
+                CORINFO_SIG_INFO calleeSigInfo;
+                _compiler->eeGetMethodSig(callNode->gtCallMethHnd, &calleeSigInfo);
+
+                if (needsReturnStackSlot(calleeSigInfo.retType, calleeSigInfo.retTypeClass))
+                {
+                    // replace the "CALL ref" with a "CALL void" that takes a return address as the first argument
+                    var_types callReturnType = callNode->TypeGet();
+                    GenTreeLclVar* shadowStackVar = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
+                    GenTreeIntCon* offset = _compiler->gtNewIconNode(_shadowStackLocalsSize, TYP_I_IMPL);
+                    GenTree* returnValueAddress = _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, shadowStackVar, offset);
+
+                    // create temp for the return address
+                    unsigned returnTempNum = _compiler->lvaGrabTemp(true, "return value address");
```

Comment:
```suggestion
                    unsigned returnTempNum = _compiler->lvaGrabTemp(false DEBUGARG("return value address"));
```
The heuristic here is "do we use the local in multiple blocks?". But actually it does not matter.